internal/controllers/nmc_reconciler.go

```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
 		statusLabels[label] = module.Config
 	}
+	return statusLabels
+}
 
-	unloaded := make([]types.NamespacedName, 0, len(nodeModuleReadyLabels))
-	loaded := make([]types.NamespacedName, 0, len(specLabels))
+func (h *nmcReconcilerHelperImpl) PrepareAndUpdateNodeLabels(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig, node *v1.Node) ([]types.NamespacedName, []types.NamespacedName, error) {
 
-	patchFrom := client.MergeFrom(node.DeepCopy())
+	// get all the kernel module ready labels of the node
+	nodeModuleReadyLabels := sets.New[types.NamespacedName]()
+	deprecatedNodeModuleReadyLabels := sets.New[string]()
 
-	// label in node but not in spec or status - should be removed
-	for nsn := range nodeModuleReadyLabels {
-		_, inSpec := specLabels[nsn]
-		_, inStatus := statusLabels[nsn]
-		if !inSpec && !inStatus {
-			meta.RemoveLabel(
-				&node,
-				utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
-			)
+	nodeModuleReadyLabels = h.lph.GetKernelModuleReadyLabels(nodeModuleReadyLabels, *node)
+	deprecatedNodeModuleReadyLabels = h.lph.GetDeprecatedKernelModuleReadyLabels(deprecatedNodeModuleReadyLabels, *node)
 
-			unloaded = append(unloaded, nsn)
-		}
-	}
+	// get spec labels and their config
+	specLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
+	specLabels = h.lph.GetSpecLabelsAndTheirConfigs(specLabels, nmc)
 
-	// v1 ready labels, deprecated - should be removed
-	for label := range deprecatedNodeModuleReadyLabels {
-		meta.RemoveLabel(&node, label)
-	}
+	// get status labels and their config
+	statusLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
+	statusLabels = h.lph.GetStatusLabelsAndTheirConfigs(statusLabels, nmc)
+
+	unloaded := make([]types.NamespacedName, 0, len(nodeModuleReadyLabels))
+	loaded := make([]types.NamespacedName, 0, len(specLabels))
+
+	// label in node but not in spec or status - should be removed
+	unloaded = h.removeOrphanedLabels(nodeModuleReadyLabels, specLabels, statusLabels, unloaded)
 
 	// label in spec and status and config equal - should be added
-	for nsn, specConfig := range specLabels {
-		statusConfig, ok := statusLabels[nsn]
-		if ok && reflect.DeepEqual(specConfig, statusConfig) && !nodeModuleReadyLabels.Has(nsn) {
-			meta.SetLabel(
-				&node,
-				utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
-				"",
-			)
+	loaded = h.addEqualLabels(nodeModuleReadyLabels, specLabels, statusLabels, loaded)
 
-			loaded = append(loaded, nsn)
-		}
+	if err := h.nodeAPI.UpdateNodeLabels(ctx, deprecatedNodeModuleReadyLabels, loaded, unloaded, node); err != nil {
+		return nil, nil, fmt.Errorf("could not patch node: %v", err)
 	}
 
-	if err := h.client.Patch(ctx, &node, patchFrom); err != nil {
-		return fmt.Errorf("could not patch node: %v", err)
-	}
+	return loaded, unloaded, nil
+}
 
-	for _, nsn := range unloaded {
+func (h *nmcReconcilerHelperImpl) RecordEvents(loaded []types.NamespacedName, unloaded []types.NamespacedName, node *v1.Node) {
```

Comment:
Can we use `loadedModules` and `unloadedModules`?

Reply:
yes.
also, do you think `loadedModules` and `unloadedModules` should be []string instead?